.github/workflows/main.yml

```diff
@@ -39,25 +39,22 @@ jobs:
             ${{ runner.os }}-${{ matrix.python-version }}-pip-tests-
             ${{ runner.os }}-${{ matrix.python-version }}-pip-
             ${{ runner.os }}-${{ matrix.python-version }}
+      - name: Stestr cache
+        uses: actions/cache@v3
+        with:
+          path: |
+            .stestr/0
+            .stestr/f*
+            .stestr/m*
+            .stestr/n*
+            .stestr/t*
+          key: ${{ runner.os }}-${{ matrix.python-version }}-stestr-tests-${{ hashFiles('setup.py','requirements.txt','requirements-extras.txt','requirements-dev.txt','constraints.txt') }}
```

Comment:
The numbered files in the `.stestr` directory are the full result stream (in the [subunit format](https://github.com/testing-cabal/subunit#the-protocol)) from the previous test runs. At the end of a test run it uses the next integer as the filename. For timing purposes there is a collection of `times.dbm.*`  files which actually contain the timing data from the most recent run, it's just a key value store of test id keys to floats of elapsed time for the value. For scheduling/balancing the times [dbm](https://en.wikipedia.org/wiki/DBM_(computing)) files are all that's needed.

For terra's CI we wipe the numbered files to avoid the growth of the cache size: https://github.com/Qiskit/qiskit/blob/main/.azure/test-linux.yml#L137 before the cache action collects the files. But I think what we're doing in terra may not be working correctly (also honestly we should just change that line to `stestr history remove` it predates the `history` subcommand iirc)  as things still appear to be running in alphabetical order in terra's ci.

Reply:
There is a brief description of the repository directory's contents in the docs here: https://stestr.readthedocs.io/en/latest/MANUAL.html#repositories

Reply:
What does caching `.stestr` do that helps CI? Is it just that it saves writing some default files that happen to be slow to write in some CI systems? Should we not cache `.stestr/0`? Or just cache `.stestr` but run `stestr history remove all` at the end before the cache is saved?

I had thought the cache might help make test discovery faster, but I think there is not a good way to do that (since even a test file that is unchanged could import a factory function that generates tests from a dependency that was updated since the previous run).

Reply:
It's for test scheduling, if there is timing data available then stestr will partition the tests across the parallel workers using that information to try and maximize throughput. The default scheduler sorts all the tests by their duration in the timing database and will then try to pack the test workers to balance the runtime evenly and minimize runtime.

If you've been running stestr locally you can experiment with this pretty easily if you do `rm -rf .stestr && tox -epy && tox -epy`, you can observe the difference in the runtime between the first and second execution of the unit tests (and also see the order of execution should be different).

Reply:
Thanks @mtreinish, I didn't scroll down enough to see the deletion of numbered files. I tried caching only the times.* files, but I also needed to include `format` or I'd get the error "The specified repository directory ./stestr already exists. Please check if the repository already exists or select a different path". I also updated the cache name to use the run number instead of the hash of requirement files the pip cache is using.

Reply:
If I follow the documentation for the cache action and for `run_number` correctly, I don't think you need to add `run_number` to the cache key. The cache action documentation says that the cache is scoped to a branch and has read only access to the cache of the default branch. The run number is incremented every time you run a workflow for a branch (not counting re-runs). So if you push a new commit, the run number on a PR will increment, but I think you would still want the cache from the previous commit on the PR. With your restore keys, you will still be able to fall back to the last commit's cache, but I don't see a benefit to keying on the number. In either case, you will use the cache from the previous commit when you push and will reuse the current commit's cache if you re-run. There is not much downside to having the build number other than that it will keep extra old runs in the cache.